learningequality/studio
# Feature flag support

## Summary
Allows for customization of Studio behavior on per-user basis through features defined in a static JSON file and granted through the admin interface.

### Description of the change(s) you made
- Adds `feature_flags` JSONB column to the User model
- Adds `jsonschema` dependency and resolves cross-dependency versioning conflicts
- Adds JSON file in form of a JSON schema to itemize available feature flags and validate feature flag data prior to saving
- Adds `static` alias to Webpack for loading JSON schema on frontend
- Exposes table of feature flags using the JSON schema in the Studio admin UI for toggling the features
- Refactors older user details API view into the `AdminUserViewSet`
- Fixes/refactors permissions for editing user record
- Adds function to garbage collect that cleans up lingering feature flags in the database

### Manual verification steps performed
1. Log into Studio admin
2. Click "Users" tab
3. View a user's details
4. Enable the dev feature
5. Verify the feature's state for the user in the database
6. Disable the dev feature
7. Verify the feature's state for the user in the database
8. Manually add non-existent feature flag to the user record in the database
9. Run `python manage.py garbage_collect` 
10. Verify the user in the database no longer has the feature flag manually added

### Screenshots (if applicable)
![Screenshot from 2021-05-03 11-33-39](https://user-images.githubusercontent.com/819838/116917564-ab139a00-ac03-11eb-95d1-8137f3f62ed9.png)

| Enabling | Disabling |
| --- | --- |
| ![Screenshot from 2021-05-03 11-31-15](https://user-images.githubusercontent.com/819838/116917448-891a1780-ac03-11eb-839a-3b9eff27182f.png) | ![Screenshot from 2021-05-03 11-30-55](https://user-images.githubusercontent.com/819838/116917470-90412580-ac03-11eb-8d54-653e4ba0fd4e.png) |


## References

Resolves https://github.com/learningequality/studio/issues/3120

## Comments
In the above issue it has the following deliverable:
> Add under setting disk space section, to allow reuse of Save Changes button already present

I did not follow the above deliverable because the save button is contained in a standalone component which would either require additional refactoring or would otherwise not be nominally understandable. Instead I added it below and used a `KSwitch` which is expected to apply the change immediately (without the need for a save button).

----

### Contributor's Checklist


PR process:

Studio-specifc:



Testing:

___

### Reviewer's Checklist
#### This section is for reviewers to fill out.
